### DISCUSSION TITLE
Guidance in using the Editing Scope

### DISCUSSION THREAD

**Discussion by arbrumbaugh** [iTwin/itwinjs-core#7043] [Q&A]
Our current implementation for entering and exiting the editing scope is based on view events.

When the onChangeView() event fires we change the editing scope. For example, when switching between a 3d and a 2d view, we enable (exit/enter) the Editing Scope. 

We regularly see, "Error: Cannot exit editing scope after it is disconnected from the iModel", when switching between views or frontstages.

- When switching from 3d --> 2d view
  - The possible cause is the [DrawingNavigationAid](https://github.com/iTwin/appui/blob/master/ui/imodel-components-react/src/imodel-components-react/navigationaids/DrawingNavigationAid.tsx#L116) component in our 2d view
- When switching from our Home frontstage (non-graphical) --> iModel Editor frontstage (editing geometry)
  - The possible cause is the onChangeView() event is triggered many times, causing our code to be called multiple times (calling to `exit` the editing scope more than once) 

We noticed there is an `internal` property [isDisposed](https://github.com/iTwin/itwinjs-core/blob/master/core/frontend/src/GraphicalEditingScope.ts#L141) that we could potentially use to solve our issue. Are there plans to make that `public` at any point?

Mainly, we want to enable (exit/enter) the editing scope when switching between our (potentially many) geometric models.
Is there a recommended event we should exit/enter the editing scope? Or maybe we should manually handle this ourselves when switching between our geometric models?

Here is the code for our current implementation for enabling the editing scope:

```
const enterEditingScope = async (iModel: IModelConnection): Promise<void> => {
  if (!iModel.isBriefcaseConnection()) return;

  if (iModel.editingScope) return;

  const supportsGraphicalEditing = await iModel.supportsGraphicalEditing();
  if (!supportsGraphicalEditing) return;

  await iModel.enterEditingScope();
};

const exitEditingScope = async (iModel: IModelConnection): Promise<void> => {
  if (!iModel.isBriefcaseConnection()) return;

  if (!iModel.editingScope) return;

  await iModel.editingScope.exit();
};

/**
 * Handle entering and exiting the editing scope in the iModel
 */
export const useEditingScope = (): void => {
  const enableEditingScope = async (vp: Viewport) => {
    const iModel = vp.iModel;
    await exitEditingScope(iModel);
    await enterEditingScope(iModel);
  };

  IModelApp.viewManager.onSelectedViewportChanged.addListener(async (args: SelectedViewportChangedArgs) => {
    if (args.current) {
      // Don't add the listener if same view
      if (args.current.view.id === args.previous?.view.id)
        return;

      args.current.onChangeView.addListener(async (vp: Viewport) => {
        await enableEditingScope(vp);
      });
    } else {
      if (args.previous)
        await exitEditingScope(args.previous.iModel);
    }
  });

  IModelApp.viewManager.onViewOpen.addListener(async (vp: ScreenViewport) => {
    await enableEditingScope(vp);
  });
};
```

Versions:
- iTwin.js core: 4.7.0
- Mango: 1.0.54

**Comment by arbrumbaugh** [comment]
@matthaleky @rajashanraja @AparupaB

**Comment by pmconne** [comment]
> We noticed there is an internal property [isDisposed](https://github.com/iTwin/itwinjs-core/blob/master/core/frontend/src/GraphicalEditingScope.ts#L141) that we could potentially use to solve our issue. Are there plans to make that public at any point?

I would be interested to know what results you observe if you check that property before exiting the scope.

I'd also be interested to see the webpacked (or otherwise bundled-for-production) JavaScript corresponding to the TypeScript snippet you posted. Bundlers do "optimizations" around property getters, where they assume repeated calls to a getter (even after an `await`) will always return the same object, and so elide them. That could explain how you're invoking `exit` on a scope that is no longer connected to the iModel.

> Mainly, we want to enable (exit/enter) the editing scope when switching between our (potentially many) geometric models.
Is there a recommended event we should exit/enter the editing scope? Or maybe we should manually handle this ourselves when switching between our geometric models?

At the end of the day we have to leave it up to the app to decide when it's an appropriate time to enter or exit an editing scope.
Switching between "active" models or front-stages seem like the kinds of context switches where we might expect the user to have completed a bunch of editing operations. Or perhaps you've got a tool that places or modifies a large number of elements at once, which might make sense to do within a single scope.

**Reply by arbrumbaugh** [reply]
I added a check for `isDisposed` in `exitEditingScope()` :

```
const exitEditingScope = async (iModel: IModelConnection): Promise<void> => {
  if (!iModel.isBriefcaseConnection())
    return;

  if (!iModel.editingScope)
    return;

  if (iModel.editingScope.isDisposed)
    return;

  try {
    await iModel.editingScope.exit();
    Logger.logInfo(AppLoggerCategory.Frontend, `Successfully exited editing scope.`);
  } catch (error: any) {
    Logger.logError(AppLoggerCategory.Frontend, `Error attempting to exit editing scope: ${error.message}`);
  }
};
```

With this, we no longer get the error "Error: Cannot exit editing scope after it is disconnected from the iModel". That seems to have resolved our issue, even though it may not be the appropriate solution.

I noticed that after initially calling `...editingScope.exit()`, the editing scope would be exited successfully, with the editing scope internal property `_isDisposed` set to `true` and the `onExiting` and `onExited` events raised (see [GraphicalEditingScope.exit()](https://github.com/iTwin/itwinjs-core/blob/master/core/frontend/src/GraphicalEditingScope.ts#L112)). 

With the subsequent calls to `...editingScope.exit()`, the `_editingScope` had not been set to `undefined` yet in the [GraphicalEditingScope.onExited()](https://github.com/iTwin/itwinjs-core/blob/master/core/frontend/src/BriefcaseConnection.ts#L152C1-L158C1) event listener inside the `ModelChangeMonitor` class, but the editing scope itself had been disposed, with `isDisposed === true`.

So, checking the `isDisposed` property resolved our issue but since it's `internal` we'll need to find another solution.
